ir/instr.cpp

```diff
@@ -170,184 +170,279 @@ static expr fm_poison(const expr &a, const expr &b, const expr &val,
-    vector<StateValue> vals = { { a - b, expr(not_poison) } };
-    vals.emplace_back((!a.sub_no_uoverflow(b)).toBVBool(), move(not_poison));
-    return getType().getAsAggregateType()->aggregateVals(vals);
-  }
-  case SMul_Overflow: {
-    vector<StateValue> vals = { { a * b, expr(not_poison) } };
-    vals.emplace_back((!a.mul_no_soverflow(b)).toBVBool(), move(not_poison));
-    return getType().getAsAggregateType()->aggregateVals(vals);
-  }
-  case UMul_Overflow: {
-    vector<StateValue> vals = { { a * b, expr(not_poison) } };
-    vals.emplace_back((!a.mul_no_uoverflow(b)).toBVBool(), move(not_poison));
-    return getType().getAsAggregateType()->aggregateVals(vals);
-  }
   case FAdd:
-    val = a.fadd(b);
-    not_poison &= !fm_poison(a, b, val, flags);
+    scalar_op = [&](auto a, auto b, auto &t) -> StateValue {
+      auto non_poison = a.non_poison && b.non_poison;
+      non_poison &= !fm_poison(a.value, b.value, a.value.fadd(b.value), flags);
+      return { a.value.fadd(b.value), move(non_poison) };
+    };
     break;
 
   case FSub:
-    val = a.fsub(b);
-    not_poison &= !fm_poison(a, b, val, flags);
+    scalar_op = [&](auto a, auto b, auto &t) -> StateValue {
+      auto non_poison = a.non_poison && b.non_poison;
+      non_poison &= !fm_poison(a.value, b.value, a.value.fsub(b.value), flags);
+      return { a.value.fsub(b.value), move(non_poison) };
+    };
     break;
 
   case FMul:
-    val = a.fmul(b);
-    not_poison &= !fm_poison(a, b, val, flags);
+    scalar_op = [&](auto a, auto b, auto &t) -> StateValue {
+      auto non_poison = a.non_poison && b.non_poison;
+      non_poison &= !fm_poison(a.value, b.value, a.value.fmul(b.value), flags);
+      return { a.value.fmul(b.value), move(non_poison) };
+    };
     break;
 
   case FDiv:
-    val = a.fdiv(b);
-    not_poison &= !fm_poison(a, b, val, flags);
+    scalar_op = [&](auto a, auto b, auto &t) -> StateValue {
+      auto non_poison = a.non_poison && b.non_poison;
+      non_poison &= !fm_poison(a.value, b.value, a.value.fdiv(b.value), flags);
+      return { a.value.fdiv(b.value), move(non_poison) };
+    };
     break;
 
   case FRem:
-    // TODO; Z3 has no support for LLVM's frem which is actually an fmod
-    not_poison = !fm_poison(a, b, val, flags);
+    scalar_op = [&](auto a, auto b, auto &t) -> StateValue {
+      auto non_poison = a.non_poison && b.non_poison;
+      non_poison &= !fm_poison(a.value, b.value, expr(), flags);
+      // TODO; Z3 has no support for LLVM's frem which is actually an fmod
+      return { expr(), move(non_poison) };
+    };
     break;
   }
 
-  return { move(val), move(not_poison) };
+  if (getType().isVectorType()) {
+    StateValue v;
+    auto aty = getType().getAsAggregateType();
+    for (unsigned i = 0 ; i < aty->numElementsConst(); ++i) {
+      auto a_i = aty->extract(s[*lhs], i);
+      auto b_i = aty->extract(s[*rhs], i);
+      auto &cty = aty->getChild(i);
+      auto vv = cty.toBV(scalar_op(a_i, b_i, cty));
```

Comment:
pls do a move(a_i), move(b_i) here.

Reply:
actually, I think the signature of the function needs to take `const StateValue &` instead because that's the return value of state[].  Right now it's always calling the copy constructor even for the scalar case.